src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    /// <summary>
+    ///     Hosts a <see cref="IWorkspaceProjectContext"/> and handles the interaction between the project system and the language service.
+    /// </summary>
+    internal partial class LanguageServiceHost : OnceInitializedOnceDisposedAsync
+    {
+        private readonly IUnconfiguredProjectCommonServices _commonServices;
+        private readonly Lazy<IProjectContextProvider> _contextProvider;
+        private readonly IProjectAsynchronousTasksService _tasksService;
+        private readonly IActiveConfiguredProjectSubscriptionService _activeConfiguredProjectSubscriptionService;
+        private IWorkspaceProjectContext _projectContext;
+        private IDisposable _evaluationSubscriptionLink;
+        private IDisposable _designTimeBuildSubscriptionLink;
+
+        [ImportingConstructor]
+        public LanguageServiceHost(IUnconfiguredProjectCommonServices commonServices,
+                                   Lazy<IProjectContextProvider> contextProvider,
+                                   [Import(ExportContractNames.Scopes.UnconfiguredProject)]IProjectAsynchronousTasksService tasksService,
+                                   IActiveConfiguredProjectSubscriptionService activeConfiguredProjectSubscriptionService)
+
+            : base(commonServices.ThreadingService.JoinableTaskContext)
+        {
+            Requires.NotNull(contextProvider, nameof(contextProvider));
+            Requires.NotNull(tasksService, nameof(tasksService));
+            Requires.NotNull(activeConfiguredProjectSubscriptionService, nameof(activeConfiguredProjectSubscriptionService));
+
+            _commonServices = commonServices;
+            _contextProvider = contextProvider;
+            _tasksService = tasksService;
+            _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+
+            Handlers = new OrderPrecedenceImportCollection<ILanguageServiceRuleHandler>(projectCapabilityCheckProvider: commonServices.Project);
+        }
+
+        [ImportMany]
+        public OrderPrecedenceImportCollection<ILanguageServiceRuleHandler> Handlers
+        {
+            get;
+        }
+
+        [ProjectAutoLoad(ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService2)]
+        private Task OnProjectFactoryCompletedAsync()
+        {
+            using (_tasksService.LoadedProject())
```

Comment:
Still playing around with initialization around this - debating whether to wait for first evaluation.